---
title: "03 Exploratory Analysis"
author: "Roxana Andreea Moldovan Moldovan"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: false
    toc_float: false
    number_sections: false 
    code_folding: hide
    theme: flatly
    bg: '#FFFFFF'
    fg: '#202020'
    primary: '#90C432'
    secondary: '#2494b5' 
  pdf_document:
    extra_dependencies: "subfig"
linkcolor: blue
urlcolor: blue
citecolor: blue
header-includes:
- \usepackage{subfig}
- \usepackage{float}
fig_caption: yes 
link-citations: TRUE

---


```{r setup, include=FALSE}
centrado = "center"
knitr::opts_chunk$set(echo = TRUE)
doAll = T
loadData = TRUE
doBP = F
doPCA = F
doClus = F
doBar = F
doBatch = F
doShow = T
doAnot = T
doAnot2 = T
doBEC = F
doRes = F


doBP = T
doPCA = T
doClus = T
doBar = T
doBatch = T
doShow = T
doAnot = T
doAnot2 = T
doBEC = T
doRes = T
```  

```{r echo=T, eval=TRUE, message=FALSE, warning=FALSE}
library(pacman)
pacman::p_load(affy)
pacman::p_load(DT)
pacman::p_load(ggdendro)
pacman::p_load(ggplot2)
pacman::p_load(ggpubr)
pacman::p_load(ggnewscale)
pacman::p_load(glue) 
pacman::p_load(ggrepel)
pacman::p_load(SummarizedExperiment)
pacman::p_load(sva)
pacman::p_load(flextable)
pacman::p_load(officer)
pacman::p_load(dplyr)
pacman::p_load(ComplexHeatmap)
pacman::p_load(ComplexUpset)
pacman::p_load(plyr)
pacman::p_load(stringr)
pacman::p_load(tidyr)

## Parametros
DirBase = "./"
DirData = glue("{DirBase}/Data")
DirPlots =  glue("{DirBase}/Plots")

## Carga de funciones
source(glue("Functions/Functions.R"))

colors = c("C" = "#90C432", "T2D" = "#884EA0" , "Ob"= "#2494B5") #, "Ob_IR" = "#D8609A")
alpha = c("M" = 0.95, "F" = 0.45)
alpha0 = c("M" = 0.65, "F" = 0.25)
forma = c(16,17)
```


#    {.tabset .tabset-fade -}  

&nbsp;   
&nbsp;  

## General information {.tabset .tabset-fade -}   

&nbsp;  

This script allows you to explore and display the data from the selected studies. 
selected. 
The raw data were downloaded from GEO and ArrayExpress, the image data from miroarrays were converted to unnormalized counts. 
The image data from the microarrays were converted to unnormalized counts.   

Given the variability of the information, the studies were unified according to the following criteria, taking into account these characteristics.
the following criteria, taking into account these characteristics and the levels present at the time 
and the levels present at that time, depending on the experiment:   

* *Sample identifier*:   
Name assigned to the sample according to a simple ascending numbering.  

```
- Sample:  

Levels : S1...Sn
```  

* *Diabetes*:   
Medical condition that categorizes the patient based on response to insulin and and the presence of T2D   

```
- Diabetes: 

  · IS: “Insulin Sensitive” (Insulin Sensitive or control)
  · IR: “Insulin Resistance” (insulin resistant or diabetic)

Levels : IS, IR
```   

* *Obesity*:   
Condition that categorizes the patient in terms of obesity. It is defined according to BMI; for this study, a limit of 30 kg/m2 has been set in all cases, based on previously defined and documented medical criteria.    

```
- Obesity: 

  · Np: Normal weigth (BMI > 30kg/m2)
  · Ob: Obesity (BMI >= 30kg/m2)

Levels : Np, Ob
```   

* *Group*:  
It is defined as the combination of the patient's obesity status and T2D, which facilitates the analysis and comparison of different subgroups within the patient. In this analysis, three levels are defined, obtained by the linear combination of the two previous variables and distributed among the different studies.  These have been recoded to correspond to the nomenclature of the paper.  

```
- Group:  

  · C: Control (Np_IS)
  · Ob: Obesity (Ob_IS) 
  · T2D: Type 2 diabetes (Ob_IR)

Levels : C, Ob, T2D
```    

* *Sex*:    
Condition representing the sex of the patient from which the sample originated.       

```
- Sex: 

  · M: "Male"
  · F: "Female"

Levels : M, F
```    

* *Tissue*:    
Detailed classification of the tissue of origin of the sample. According to the origin of the sample and the previously described clinical and physiological differences between the different tissues and their disposition, two clinically relevant groups have been defined: SAT and VAT.

```
- Tissue: 

  · SAT: Subcutaneous Adipose Tissue 
  · VAT: Visceral Adipose Tissue 

Levels : SAT, VAT
```   

* *Age Group*:    
Classification of patients by age group. Because of the prevalence and importance of DM2 in advanced age and the metabolic and hormonal differences typical of age and development, patients were divided into two groups. 

```
- AgeGroup: 

  · Adult: Adult
  · Non adult: Children under the age of 18

Levels : Adult, Non adult
```    

* *Patient identifier*:    
Identifier code assigned to each sample according to the patient of origin. In case two or more samples come from the same patient, they have the same identifier.      

```
- Patient:  

Levels : P1...Pn
```    

After unifying the information from the studies, they have been processed and stored in objects of the SummarizedExperiment and ExpressionSet classes.     

&nbsp;  

***      

&nbsp;   



## Exploratory analysis  {.tabset .tabset-fade -}   


### E-MEXP-1425 {.tabset .tabset-fade -}    

&nbsp;   

Transcriptomic study using arrays on the Affymetrix platform. 
GeneChip Human Genome U133 Plus 2.0 [HG-U133_Plus_2].  

The aim of this study was to investigate the global transcriptomic profiles of twin pairs with discordant body mass indexes. These profiles were analyzed on the basis of abdominal subcutaneous adipose tissue (SAT) and also provide information on BMI, allowing patients to be classified into obese and controls on this basis. 


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "E_MEXP_1425"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data 
E_MEXP_1425_rawData = rawData = get(load(glue("{DirRData}/rawData.RData")))
E_MEXP_1425_normData = normData = get(load(glue("{DirRData}/normData2.RData")))
E_MEXP_1425_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)] 
```  

&nbsp;

#### Summary  {-}     


```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the E_MEXP_1425 study by group and sex._"}  

BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.5 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 
```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the E_MEXP_1425 study by group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  


#### Raw Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(rawData) == "SummarizedExperiment",
                       assay(rawData),
                       exprs(rawData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}rawBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(rawData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(rawData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(rawData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.4,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}rawHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    



#### Normalized Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.4,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normHC.svg"))

``` 

&nbsp;   

***      

&nbsp;     

#### Final data (normalized and annotated) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.4,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study E-MEXP-1425, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


### GSE2508 {.tabset .tabset-fade -}    

&nbsp;   

Transcriptomic study by arrays using the platforms:  
* [HG_U95A] Affymetrix Human Genome U95A Array.  
* HG_U95B] Affymetrix Human Genome U95B Array.  
* HG_U95C] Affymetrix Human Genome U95C Array.  
* HG_U95D] Affymetrix U95D Human Genome Array.  
* HG_U95E] Affymetrix U95E Human Genome Array.  
* HG_U95Av2] Affymetrix Human Genome U95 version 2 array.  

This study aims to study the differential transcriptomic profiles between healthy, non-insulin resistant patients, 20 obese and 19 with normal weight from adipocytes derived from abdominal subcutaneous adipose tissue, Indians, considering obese subjects those with a BMI greater than or equal to 30kg/m2. 


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "GSE2508"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data 
GSE2508_rawData = GSE2508_normData = normData= get(load(glue("{DirRData}/normData.RData")))
GSE2508_normData2 = normData2= get(load(glue("{DirRData}/normData2.RData")))
GSE2508_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)]  
```  


&nbsp;    

#### Summary  {-}     



```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE2508 per group and sex._"}  

BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.5 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 

```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE2508 per group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  


#### Normalized Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  



##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.25,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Batch effect {.tabset .tabset-fade -}   

We study the possible variables that produce batch effect.  


```{r echo=T, eval=doBatch, message=FALSE, warning=FALSE}
# Create the variable indicating whether they are new or reanalyzed samples  
batchVar = phenoData$Batch 
ColorList = c("#00AFBB", "#CB326D")
ColorPal <- colorRampPalette(ColorList)
colorsBath <- ColorPal(length(levels(batchVar)))
```


&nbsp;       


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, attending to the batch effect._"}
### PCA
# Computing PCA 
PCAplot = RX_multiplotPCA(normData,
                          multiplot = FALSE, 
                          labels = "Sample",
                          colorColumn = "Batch",
                          shapeColumn = "Sex",
                          shapes = forma,
                          colors = colorsBath,
                          colorLegend = "Batch",
                          shapeLegend = "Sexo",
                          alpha = 0.8) 

# Saving...
ggsave(PCAplot, device = "svg",
       filename = glue("{DirPlot}/{Acc}BatchPCA0.svg"),
       height = 6, width = 8) 
  
PCAplot
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, attending to the batch effect._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}BatchPCA0.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Batch",
                            colorPal = colorsBath,
                            colorOutColumn = "Batch",
                            colorOutPal = colorsBath,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Batch",
                         colorPal = colorsBath,
                         colorOutColumn = "Batch",
                         colorOutPal = colorsBath,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.25,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}BatchHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"} 

knitr::include_graphics(glue("{DirPlot}/{Acc}BatchHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Corrección del batch effect {.tabset .tabset-fade -}   

Corregimos en batch effect. 

&nbsp;    

##### Boxplot {-}   


```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData2) == "SummarizedExperiment",
                       assay(normData2),
                       exprs(normData2))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normNoBatchBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```     


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Análsisis de comopnentes principales del estudio GSE2508,attending to the batch effect._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchBoxPlot.svg"))

``` 



##### PCA {.tabset .tabset-fade -}     
  
###### Exp var {-}    

```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData2,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo")
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
} else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData2,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData2,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                   common.legend = TRUE,
                                   legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 


    
```   

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchPCA.svg"))

``` 

&nbsp;   

***      

&nbsp;   


###### Batch var {-}     

```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Análsisis de comopnentes principales del estudio GSE2508,attending to the batch effect._"}
### PCA
# Computing PCA 
PCAplot = RX_multiplotPCA(normData2,
                          multiplot = FALSE,
                          labels = "Sample", 
                          colorColumn = "Batch",
                          shapeColumn = "Sex",
                          shapes = forma,
                          colors = colorsBath,
                          colorLegend = "Batch",
                          shapeLegend = "Sexo",
                          alpha = 0.8) 

# Saving...
ggsave(PCAplot, device = "svg",
       filename = glue("{DirPlot}/{Acc}BatchCorrPCA0.svg"),
       height = 6, width = 8) 
  
PCAplot
    
```    

&nbsp;    

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Análsisis de comopnentes principales del estudio GSE2508,attending to the batch effect._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchCorrPCA0.svg"))

```  


&nbsp;    

***     

&nbsp;  


##### Clustering  {.tabset .tabset-fade -}      

###### Exp var {-} 

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"} 
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance

EU_dist = RX_multiplotDendo(normData2,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            )  

# Correlation
Corr = RX_multiplotDendo(normData2,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )   

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.25,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normNoBatchHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"} 
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchHC.svg"))

```

&nbsp;   

***      

&nbsp;   


###### Batch var {-}     


```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"} 
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData2,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Batch",
                            colorPal = colorsBath,
                            colorOutColumn = "Batch",
                            colorOutPal = colorsBath,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            )  

# Correlation
Corr = RX_multiplotDendo(normData2,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Batch",
                         colorPal = colorsBath,
                         colorOutColumn = "Batch",
                         colorOutPal = colorsBath,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )   

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.25,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}BatchCorrHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```     

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9  , fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"} 
  
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchCorrHC.svg"))

```


&nbsp;   

***      

&nbsp; 


#### Final data (normalized and annotated) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.25,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=9, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE2508, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    



### GSE20950 {.tabset .tabset-fade -}    

&nbsp;  

Transcriptomic study by means of arrays using the [HG-U133_Plus_2] Affymetrix Human Genome U133 Plus 2.0 Array (GPL570) platform.  

This study highlights Obesity as a risk factor in the development of diseases and metabolic disorders, more precisely in the development of insulin resistance, for which the Expression profiles of obese insulin resistant and insulin sensitive patients have been analyzed from samples of subcutaneous abdominal adipose tissue (SAT) and omental (VAT) of 20 adult patients, with an average age of 42 years.   


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "GSE20950"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data  
GSE20950_normData = normData= get(load(glue("{DirRData}/normData.RData")))
GSE20950_normData2 = normData2= get(load(glue("{DirRData}/normData2.RData")))
GSE20950_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)] 

```  


&nbsp;    

#### Summary  {-}     


```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=10, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE20950 per group and sex._"}  

BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.5 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 

```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=10, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE20950 per group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  


#### Normalized Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    



```{r echo=F, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha,
                     dir = "v",
                     ncol = 2)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlotTFM.svg"),
       height =  4, #2 + (3*round(nTis/2)),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
#BoxPlot
```   

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.50,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Batch effect {.tabset .tabset-fade -}   

We study the possible variables that produce batch effect.  


```{r echo=T, eval=doBatch, message=FALSE, warning=FALSE}
# Create the variable indicating whether they are new or reanalyzed samples  
batchVar = phenoData$Batch 
ColorList = c("#00AFBB", "#CB326D")
ColorPal <- colorRampPalette(ColorList)
colorsBath <- ColorPal(length(levels(batchVar)))
```


&nbsp;       


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, attending to the batch effect._"}
### PCA
# Computing PCA 
PCAplot = RX_multiplotPCA(normData,
                          multiplot = FALSE, 
                          labels = "Sample",
                          colorColumn = "Batch",
                          shapeColumn = "Sex",
                          shapes = forma,
                          colors = colorsBath,
                          colorLegend = "Batch",
                          shapeLegend = "Sexo",
                          alpha = 0.8) 

# Saving...
ggsave(PCAplot, device = "svg",
       filename = glue("{DirPlot}/{Acc}BatchPCA0.svg"),
       height = 6, width = 8) 
  
PCAplot
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, attending to the batch effect._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}BatchPCA0.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Batch",
                            colorPal = colorsBath,
                            colorOutColumn = "Batch",
                            colorOutPal = colorsBath,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Batch",
                         colorPal = colorsBath,
                         colorOutColumn = "Batch",
                         colorOutPal = colorsBath,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.50,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}BatchHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"}
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Corrección del batch effect {.tabset .tabset-fade -}   

Corregimos en batch effect. 

&nbsp;    

##### Boxplot {-}   


```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData2) == "SummarizedExperiment",
                       assay(normData2),
                       exprs(normData2))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normNoBatchBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```     


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Análsisis de componentes principales del estudio GSE20950,attending to the batch effect._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchBoxPlot.svg"))

``` 



##### PCA {.tabset .tabset-fade -}     
  
###### Exp var {-}    

```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData2,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo")
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
} else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData2,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData2,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                   common.legend = TRUE,
                                   legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
  } 
    
```   

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchPCA.svg"))

``` 

&nbsp;   

***      

&nbsp;   


###### Batch var {-}     

```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Análsisis de componentes principales del estudio GSE20950,attending to the batch effect._"}
### PCA
# Computing PCA 
PCAplot = RX_multiplotPCA(normData2,
                          multiplot = FALSE,
                          labels = "Sample",
                          colorColumn = "Batch",
                          shapeColumn = "Sex",
                          shapes = forma,
                          colors = colorsBath,
                          colorLegend = "Batch",
                          shapeLegend = "Sexo",
                          alpha = 0.8) 

# Saving...
ggsave(PCAplot, device = "svg",
       filename = glue("{DirPlot}/{Acc}BatchCorrPCA0.svg"),
       height = 6, width = 8) 
  
PCAplot
    
```    

&nbsp;    

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Análsisis de componentes principales del estudio GSE20950,attending to the batch effect._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchCorrPCA0.svg"))

```  


&nbsp;    

***     

&nbsp;  


##### Clustering  {.tabset .tabset-fade -}      

###### Exp var {-} 

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"} 
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance

EU_dist = RX_multiplotDendo(normData2,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            )  

# Correlation
Corr = RX_multiplotDendo(normData2,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )   

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.50,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normNoBatchHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"} 
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchHC.svg"))

```

&nbsp;   

***      

&nbsp;   


###### Batch var {-}     


```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"} 
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData2,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Batch",
                            colorPal = colorsBath,
                            colorOutColumn = "Batch",
                            colorOutPal = colorsBath,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            )  

# Correlation
Corr = RX_multiplotDendo(normData2,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Batch",
                         colorPal = colorsBath,
                         colorOutColumn = "Batch",
                         colorOutPal = colorsBath,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )   

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.50,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}BatchCorrHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```     

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"} 
  
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchCorrHC.svg"))

```


&nbsp;   

***      

&nbsp; 


#### Final data (normalized and annotated) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.50,10,0.0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=10, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE20950, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    




### GSE29718 {.tabset .tabset-fade -}    

&nbsp;  

Transcriptomic study using the [HuGene-1_0-st] Affymetrix Human Gene 1.0 ST Array [transcript (gene) version] platform (GPL6244).  

In an attempt to elucidate the genetic mechanisms behind obesity, which may subsequently lead to the development of type 2 diabetes, this study aims to study the differences between obese children, with the goal of minimizing the effects of puberty.  
For this purpose, 15 samples of abdominal subcutaneous adipose tissue (SAT) and 5 of visceral adipose tissue (VAT) were taken from children under 9 years of age.   


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "GSE29718"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data 
GSE29718_rawData = rawData = get(load(glue("{DirRData}/rawData.RData")))
GSE29718_normData = normData= get(load(glue("{DirRData}/normData.RData")))
GSE29718_normData2 = normData2= get(load(glue("{DirRData}/normData2.RData")))
GSE29718_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)] 

```  


&nbsp;    

#### Summary  {-}     



```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=10, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE29718 per group and sex._"}  

BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.5 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 

```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=10, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE29718 per group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  



#### Raw Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(rawData) == "SummarizedExperiment",
                       assay(rawData),
                       exprs(rawData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}rawBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(rawData) == "SummarizedExperiment",
                       assay(rawData),
                       exprs(rawData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 
BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha,
                     dir = "v",
                     ncol = 2)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}rawBoxPlotTFM.svg"),
       height =  4, #2 + (3*round(nTis/2)),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
#BoxPlot
```    

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(rawData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(rawData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(rawData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.7,10,0.1)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}rawHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Normalized Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    

```{r echo=F, eval=doBP, message=FALSE,warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha,
                     dir = "v",
                     ncol = 2)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlotTFM.svg"),
       height =  4, #2 + (3*round(nTis/2)),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
#BoxPlot
```   

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.7,10,0.1)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Batch effect {.tabset .tabset-fade -}   

We study the possible variables that produce batch effect.  


```{r echo=T, eval=doBatch, message=FALSE, warning=FALSE}
# Create the variable indicating the batch in which the samples have been analyzed
batchVar = phenoData$Batch 
ColorList = c("#00AFBB", "#CB326D")
ColorPal <- colorRampPalette(ColorList)
colorsBath <- ColorPal(length(levels(batchVar)))
```


&nbsp;       


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, attending to the batch effect._"}
### PCA
# Computing PCA 
PCAplot = RX_multiplotPCA(normData,
                          multiplot = FALSE, 
                          labels = "Sample",
                          colorColumn = "Batch",
                          shapeColumn = "Sex",
                          shapes = forma,
                          colors = colorsBath,
                          colorLegend = "Batch",
                          shapeLegend = "Sexo",
                          alpha = 0.8) 

# Saving...
ggsave(PCAplot, device = "svg",
       filename = glue("{DirPlot}/{Acc}BatchPCA0.svg"),
       height = 6, width = 8) 
  
PCAplot
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, attending to the batch effect._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}BatchPCA0.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Batch",
                            colorPal = colorsBath,
                            colorOutColumn = "Batch",
                            colorOutPal = colorsBath,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Batch",
                         colorPal = colorsBath,
                         colorOutColumn = "Batch",
                         colorOutPal = colorsBath,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.7,10,0.1)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}BatchHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"}
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Corrección del batch effect {.tabset .tabset-fade -}   

Corregimos en batch effect. 

&nbsp;    

##### Boxplot {-}   


```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData2) == "SummarizedExperiment",
                       assay(normData2),
                       exprs(normData2))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normNoBatchBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```     


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Análsisis de componentes principales del estudio GSE29718,attending to the batch effect._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchBoxPlot.svg"))

``` 



##### PCA {.tabset .tabset-fade -}     
  
###### Exp var {-}    

```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData2,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo")
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
} else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData2,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData2,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                   common.legend = TRUE,
                                   legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normNoBatchPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
  } 
    
```   

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchPCA.svg"))

``` 

&nbsp;   

***      

&nbsp;   


###### Batch var {-}     

```{r echo=T, eval=doBEC, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Análsisis de componentes principales del estudio GSE29718,attending to the batch effect._"}
### PCA
# Computing PCA 
PCAplot = RX_multiplotPCA(normData2,
                          multiplot = FALSE,
                          labels = "Sample",
                          colorColumn = "Batch",
                          shapeColumn = "Sex",
                          shapes = forma,
                          colors = colorsBath,
                          colorLegend = "Batch",
                          shapeLegend = "Sexo",
                          alpha = 0.8) #+
  #facet_wrap(facets = c(normData2$Tissue), ncol = 2, scales = "free")

# Saving...
ggsave(PCAplot, device = "svg",
       filename = glue("{DirPlot}/{Acc}BatchCorrPCA0.svg"),
       height = 6, width = 8) 
  
PCAplot
    
```    

&nbsp;    

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Análsisis de componentes principales del estudio GSE29718,attending to the batch effect._"}
  
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchCorrPCA0.svg"))

```  


&nbsp;    

***     

&nbsp;  


##### Clustering  {.tabset .tabset-fade -}      

###### Exp var {-} 

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"} 
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance

EU_dist = RX_multiplotDendo(normData2,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            )  

# Correlation
Corr = RX_multiplotDendo(normData2,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )   

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.7,10,0.1)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normNoBatchHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"} 
  
knitr::include_graphics(glue("{DirPlot}/{Acc}normNoBatchHC.svg"))

```

&nbsp;   

***      

&nbsp;   


###### Batch var {-}     


```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"} 
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData2,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Batch",
                            colorPal = colorsBath,
                            colorOutColumn = "Batch",
                            colorOutPal = colorsBath,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            )  

# Correlation
Corr = RX_multiplotDendo(normData2,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Batch",
                         colorPal = colorsBath,
                         colorOutColumn = "Batch",
                         colorOutPal = colorsBath,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )   

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.7,10,0.1)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}BatchCorrHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```     

```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped on the basis of their batch effect by color, while the shape denotes sex. This cluster has been carried out for each of the tissues, both on the basis of the correlation between the samples (B) and on the basis of the Euclidean distance (A)._"} 
  
knitr::include_graphics(glue("{DirPlot}/{Acc}BatchCorrHC.svg"))

```


&nbsp;   

***      

&nbsp; 


#### Final data (normalized and annotated) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=12, fig.width=12, fig.cap=" _**Figure :** Principal component analysis of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. *A.* PCA based on all study samples. *B.* Individual PCAs for each of the tissues from which the samples were taken._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.7,10,0.1)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]],
               EU_dist$Plot[[2]], Con[[2]], Corr$Plot[[2]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 2,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=7, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE29718, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


### GSE64567 {.tabset .tabset-fade -}    

&nbsp;  

Transcriptomic study through arrays using the Illumina HumanHT-12 V4.0 expression beadchip platform (GPL10558).  

In an attempt to elucidate the changes at the transcriptional level that lead to the development of obesity and the possible future association between obesity and type 2 diabetes, in this study samples were taken from the abdominal subcutaneous tissue of 64 unrelated, non-diabetic adult subjects with different body mass indexes (BMI).  
From the data collected in this study, based on BMI, and taking 30>= as a cut-off, we can separate the subjects into two groups: obese (n=35) and control (n=29).


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "GSE64567"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data  
GSE64567_normData = normData= get(load(glue("{DirRData}/normData.RData")))
GSE64567_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)] 

```  


&nbsp;    

#### Summary  {-}     



```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE64567 per group and sex._"}  


BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.5 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 

```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE64567 per group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  



#### Raw Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(rawData) == "SummarizedExperiment",
                       assay(rawData),
                       exprs(rawData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = "log2",
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}rawBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(rawData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=16, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(rawData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(rawData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}rawHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=16, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Normalized Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=16, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=16, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Final data (normalized and annotated) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=16, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=16, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE64567, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


### GSE92405 {.tabset .tabset-fade -}    

&nbsp;   

Transcriptomic study using the Affymetrix Human Genome U133 Plus 2.0 Array platform [CDF: Brainarray HGU133Plus2_Hs_ENTREZG_v18].  

This study aims to study the global transcriptomic profiles of twin pairs with discordant body mass indexes (more than 3kg/m2). These profiles were analyzed based on abdominal subcutaneous adipose tissue (SAT).      
  


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "GSE92405"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data 
GSE92405_rawData = rawData = get(load(glue("{DirRData}/rawData.RData")))
GSE92405_normData2 = normData2= get(load(glue("{DirRData}/normData2.RData")))
GSE92405_normData = normData= get(load(glue("{DirRData}/normData.RData")))
GSE92405_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)] 
```  


&nbsp;    

#### Summary  {-}     



```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE92405 per group and sex._"}  

BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.5 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 
```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE92405 per group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  



#### Raw Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(rawData) == "SummarizedExperiment",
                       assay(rawData),
                       exprs(rawData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}rawBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(rawData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=13, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(rawData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(rawData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}rawHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=13, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Normalized Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(normData) == "SummarizedExperiment",
                       assay(normData),
                       exprs(normData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}normBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(normData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(normData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}normPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=13, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(normData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(normData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}normHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=13, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}normHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    



#### Final data (normalized and annotated) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = NULL,
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** Boxplot of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=13, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=13, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE92405, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


### GSE141432 {.tabset .tabset-fade -}    

&nbsp;   

Transcriptomic study by high throughput sequencing.  

In this case we studied the development of type 2 diabetes in relation to obesity, for this purpose we have taken samples of abdominal subcutaneous adipose tissue from three groups of patients, healthy subjects without obesity (controls, n=9), obese healthy subjects (n=8) and obese subjects with type 2 diabetes (n=8).    
  


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "GSE141432"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data 
GSE141432_rawData = rawData = get(load(glue("{DirRData}/rawData.RData")))
GSE141432_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)] 

```  


&nbsp;    

#### Summary  {-}     



```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE141432 per group and sex._"}  


BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.6 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 

```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE141432 per group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  



#### Raw Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(rawData) == "SummarizedExperiment",
                       assay(rawData),
                       exprs(rawData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = "log2",
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}rawBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(rawData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(rawData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(rawData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.4,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}rawHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Datos finales (anotados) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = "log2",
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.4,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE141432, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


### GSE205668 {.tabset .tabset-fade -}    

&nbsp;   

Transcriptomic study by high throughput sequencing.  

The main objective is to elucidate the mechanisms that generate the UPV ('unexplained' phenotypic variation) that was observed in previous human and murine studies and that in this case, could surround Obesity. For this purpose, abdominal subcutaneous tissue samples were taken from a total of 61 samples from children under 18 years of age belonging to two experimental groups, Obesity (n=26) and control (n=35).     


```{r echo=TRUE, results = "hide", eval=loadData, message=FALSE, warning=FALSE}
# Information and parameter definition
Acc = "GSE205668"
DirrawData = paste(DirData, Acc, "01rawData", sep = "/")
DirRData = paste(DirData, Acc, "02RData", sep = "/")
DirPlot = glue("{DirPlots}/{Acc}")

# Create a directory to store the plots
system(glue("mkdir {DirPlot}"))

# Load the data 
GSE205668_rawData = rawData = get(load(glue("{DirRData}/rawData.RData")))
GSE205668_finalData = finalData = get(load(glue("{DirRData}/finalData.RData")))

# Phenotypic data
phenoData = data.frame(RX_ifelse(class(finalData) == "SummarizedExperiment",
                                colData(finalData),
                                pData(finalData)))
# Parameters
color = colors[levels(phenoData$Group)] # Colors
nTis = length(levels(phenoData$Tissue)) # Number of different tissues
nSTis = max(table(phenoData$Tissue))    # Maximun number of samples per tissue

# Color
color = colors[levels(phenoData$Group)] 

```  


&nbsp;    

#### Summary  {-}     



```{r echo=T, eval=doBar, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE205668 per group and sex._"}  

BarPlot <- RX_resumeBarPlot(phenoData,
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = FALSE,
                            alpha = alpha0,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.5 + (0.1*nTis)) +
  theme(axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  scale_x_discrete(expand = c(0.2, 0.2))

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}ResumeBarPlot.svg"),
       height = 3.5, 
       width = 4 + (2*nTis)) # Dim

BarPlot 

```  


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Summary of the number of samples comprising the study GSE205668 per group and sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}ResumeBarPlot.svg"))

```  


&nbsp;   

***      

&nbsp;  



#### Raw Data {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(rawData) == "SummarizedExperiment",
                       assay(rawData),
                       exprs(rawData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = "log2",
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}rawBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(rawData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(rawData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}rawPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=15, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(rawData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(rawData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}rawHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}rawHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    


#### Datos finales (anotados) {.tabset .tabset-fade -}    

&nbsp;    

##### Boxplot {-}    

```{r echo=T, eval=doBP, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
# Expression 
expressData =RX_ifelse(class(finalData) == "SummarizedExperiment",
                       assay(finalData),
                       exprs(finalData))

# Convert the object to the desired format
BoxPlotData = RX_dataforboxplot(expressData, 
                                transformacion = "log2",
                                phenoData = phenoData,
                                ncol = "Sample") 

BoxPlot = RX_BoxPlot(Data = BoxPlotData,
                     Fac_fill = "Group",
                     Fac_alpha = "Sex",
                     colors = colors,
                     alpha = alpha)

# Saving...
ggsave(BoxPlot, device = "svg",
       filename = glue("{DirPlot}/{Acc}finalBoxPlot.svg"),
       height =  2 + (3*nTis),
       width = RX_ifelse(nSTis < 50, 8, 12))

# Show plot
BoxPlot
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=5, fig.width=8, fig.cap=" _**Figure :** Boxplot of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalBoxPlot.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### PCA {-}  



```{r echo=T, eval=doPCA, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}
### PCA
# Computing PCA 
if(length(levels(phenoData$Tissue)) == 1){ # Global PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6, width = 8) 
  
  PCAplot
  
}  else{ # Global and per tissue PCA
  # Total PCA
  PCAplot = RX_multiplotPCA(finalData,
                            multiplot = FALSE, 
                            factor = "Tissue",
                            labels = "Sample",
                            colorColumn = "Group",
                            shapeColumn = "Sex",
                            shapes = forma,
                            colors = colors,
                            colorLegend = "Group",
                            shapeLegend = "Sexo") 
  
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA1.svg"),
         height = 6, width = 12)  
  
  # PCA per tissue
  PCAplots = RX_multiplotPCA(finalData,
                             multiplot = TRUE,
                             factor = "Tissue",
                             labels = "Sample",
                             colorColumn = "Group",
                             shapeColumn = "Sex",
                             shapes = forma,
                             colors = colors,
                             colorLegend = "Group",
                             shapeLegend = "Sexo")
  # Combining
  PCAplot2 = do.call("ggarrange", c(PCAplots,
                                    common.legend = TRUE,
                                    legend = "right"))
  # Saving... 
  ggsave(PCAplot2, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA2.svg"),
         height = 6, width = 12)  
  
  PCAplot = do.call("ggarrange", c(list(PCAplot, PCAplot2),
                                   nrow = 2,
                                   common.legend = FALSE,
                                   legend = "right",
                                   labels = "AUTO"))
  # Number of lines
  nlines = round(nTis/2) * 2
  # Saving...
  ggsave(PCAplot, device = "svg",
         filename = glue("{DirPlot}/{Acc}finalPCA.svg"),
         height = 6*nlines, width = 12) 
  
  # Show plot
  PCAplot
} 
    
```   


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=8, fig.cap=" _**Figure :** Principal component analysis of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalPCA.svg"))

``` 


&nbsp;   

***      

&nbsp;  


##### Clustering {-}       

```{r echo=T, eval=doClus, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=15, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex. This clustering has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}
### Clustering
# Hierarchical clustering and data transformation

# Euclidean distance
EU_dist = RX_multiplotDendo(finalData,
                            dir = "lr",
                            method = "euclidean",
                            factor = "Tissue",
                            colorColumn = "Group",
                            colorPal = color,
                            colorOutColumn = "Group",
                            colorOutPal = color,
                            leaves = TRUE,
                            shapeColumn = "Sex",
                            leavesShapes = c(21,22),
                            colorLegend = "none",
                            colorOutLegend = "Group",
                            shapeLegend = "Sexo",
                            multiplot = TRUE, 
                            wrap = TRUE,
                            labelSize = 3.5,
                            expandY = -0.05,
                            leavesSize = 3,
                            leavesStroke = 1,
                            expandYlbs = -0.03,
                            moveLabel = 0
                            ) 

# Correlation
Corr = RX_multiplotDendo(finalData,
                         dir = "rl",
                         method = "correlation",
                         factor = "Tissue",
                         colorColumn = "Group",
                         colorPal = color,
                         colorOutColumn = "Group",
                         colorOutPal = color,
                         leaves = TRUE,
                         shapeColumn = "Sex",
                         leavesShapes = c(21,22),
                         colorLegend = "none",
                         colorOutLegend = "Group",
                         shapeLegend = "Sexo",
                         multiplot = TRUE, 
                         wrap = TRUE,
                         labelSize = 3.5,
                         expandY = -0.05,
                         leavesSize = 3,
                         leavesStroke = 1,
                         expandYlbs = -0.03,
                         moveLabel = 0
                         )     

Con = lapply(seq(levels(phenoData$Tissue)), 
             function(x) RX_DendoRelations(EU_dist$Data[[x]],
                                           Corr$Data[[x]],
                                           size = c(0.15,10,0)))

Pltlist = list(EU_dist$Plot[[1]], Con[[1]], Corr$Plot[[1]])


DendoPlot = ggarrange(plotlist = Pltlist,common.legend = TRUE,
                      widths=c(1.0, 0.5, 1.0), ncol = 3, nrow = 1,
                      #labels = c("A","", "B"),
                      legend = "right", label.x = 0.02) 

# Saving...
ggsave(DendoPlot, device = "svg", 
       filename = glue("{DirPlot}/{Acc}finalHC.svg"),
       height = (nSTis * nTis)/4 , width = 12)

# Show plot
DendoPlot  
```    


```{r echo=T, eval=doShow, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=6, fig.width=12, fig.cap=" _**Figure :** Hierarchical clustering of the study GSE205668, samples are grouped by color based on their medical condition, with color intensity indicating sex. This cluster has been carried out both on the basis of the correlation between samples (B) and on the basis of the Euclidean distance (A)._"}  

knitr::include_graphics(glue("{DirPlot}/{Acc}finalHC.svg"))

``` 

&nbsp;   

***      

&nbsp;    



## Annotation {.tabset .tabset-fade -}   


```{r echo=T, eval=loadData, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=14}
RX_table <- function(Data,
                     Color1 = "#007FA6",
                     Color2 = "#007FA6" ,
                     Color.l = "black" ,
                     Color.l2 = "#878787",
                     Color.up = "#CB326D",
                     Color.down = "#00AFBB",
                     footer = NULL,
                     header,
                     cols,
                     cols_keys,
                     font = "Calibri" #,  hline = FALSE
                     ){
  
  ## Flextable
  # Set info
  ft <- flextable(Data, col_keys = colnames(Data))
  # Set header
  ft <- set_header_df(ft, mapping = header, key = "col_keys" )
  # Format number
  ft <- colformat_num(ft, big.mark = ".", decimal.mark = "," )
  # Set footer
  if(! is.null(footer)){
    ft <- add_footer(ft, values = footer) %>% 
      merge_h(part = "footer")  %>% 
      bold(j=1, part = "footer") %>%
      hline(border = fp_border(width = 2, color = Color1), part = "footer")
  }
  
  # Merge headers
  ft <- merge_h(ft, part = "header") %>% 
    merge_v( part = "header")
  
  # Merge contrasts
  #ft <- merge_v(ft, j = c(1,2), part = "body")
  
  # Horizontal lines
  'if (isTRUE(hline)){
    ft <- hline(ft, i=seq(from=3, to=nrow(Data), by=3),
                border = fp_border(width = 1, color = Color.l2),
                part="body")
    }'
  ft <- hline(ft,border = fp_border(width = 2, color = Color1), part = "header")
  
  # Outer bottom
  ft <- hline_bottom(ft, border = fp_border(width = 2, color = Color1))
  
  # Alignment
  ft <- flextable::align(ft, i=1, align = "center", part = "header") %>%
    flextable::align(i=2, align = "right", part = "header")
  ft <- flextable::align(ft, align = "right", part = "footer")
  ft <- flextable::align(ft, j=1, align = "left", part = "footer")
  
  
  # Header design
  ft <- fontsize(ft, i = 1:2, size = 12, part = "header") %>%
    color(i=1, color = Color2, part = "header") %>%
    color(i=2, color = Color.l2, part = "header") %>%
    bold(i=1, bold = TRUE, part = "header")
  # Rotate names
  #ft <- rotate(ft, i = 2, rotation = "btlr", part = "header", align = "bottom")
  
  # Font
  #font(ft, fontname = font, part = "all")
  
  # Matriz de Colors
  colormatrix <- ifelse(Data[, cols] == 0, Color.l2, Color.l )
  # Color
  ft <- color(ft, j=cols, color = colormatrix, part = "body") 
  # Color
  x = Data[, 1]
  colors = case_when(
    x == "Up" ~ Color.up,
    x == "Down" ~ Color.down,
    TRUE ~ Color.l
  )
  ft <- color(ft, j=1, color = colors, part = "body") %>%
    bold(j=1, bold = TRUE, part = "body")

  return(ft)
}
```


```{r echo=T, eval=doAnot, message=FALSE, warning=FALSE, fig.align = centrado, results = 'hide', out.width="98%"}

Studies = c("E_MEXP_1425",
             "GSE2508",
             "GSE20950",
             "GSE29718", 
             "GSE64567", 
             "GSE92405",
             "GSE141432",
             "GSE205668")
Platform = c("Affymetrix GeneChip Human Genome U133 Plus 2.0 [HG-U133_Plus_2]",
               "[HG_U95Av2] Affymetrix Human Genome U95 Version 2 Array
[HG_U95A] Affymetrix Human Genome U95A Array
[HG_U95B] Affymetrix Human Genome U95B Array
[HG_U95C] Affymetrix Human Genome U95C Array
[HG_U95D] Affymetrix Human Genome U95D Array
[HG_U95E] Affymetrix Human Genome U95E Array",
               "[HG-U133_Plus_2] Affymetrix Human Genome U133 Plus 2.0 Array",
               "[HuGene-1_0-st] Affymetrix Human Gene 1.0 ST Array [transcript (gene) version]",
               "Illumina HumanHT-12 V4.0 expression beadchip",
"[HG-U133_Plus_2] Affymetrix Human Genome U133 Plus 2.0 Array [CDF: Brainarray HGU133Plus2_Hs_ENTREZG_v18]",
               "Illumina NextSeq 500 (Homo sapiens)",
               "Illumina HiSeq 2500 (Homo sapiens)"
               )
Package = c("hgu133plus2.db",
            "hgu95av2.db \n
            hgu95a.db \n
            hgu95b.db \n
            hgu95c.db \n
            hgu95d.db \n
            hgu95e.db \n",
            "hgu133plus2.db",
            "hugene10sttranscriptcluster.db",
            "illuminaHumanv4.db", 
            "hgu133plus2.db",
            "org.Hs.eg.db",
            "org.Hs.eg.db") 
Origin = c("Bioconductor",
           "Bioconductor",
           "Bioconductor",
           "Bioconductor",
           "Bioconductor", 
           "Bioconductor",
           "Bioconductor",
           "Bioconductor"
           )

normDatas = sapply(Studies, function(x) ifelse(exists(glue("{x}_normData")), glue("{x}_normData"), glue("{x}_rawData")), simplify = TRUE, USE.NAMES = FALSE)
Counts0 = sapply(normDatas, function(x) nrow(get(x)))
AnotDatas = glue("{Studies}_finalData")
Counts1 = sapply(AnotDatas, function(x) nrow(get(x)))

Data = data.frame(Studies,
                  Platform,
                  Package,
                  Origin,
                  "Pre-Annotation" = Counts0,
                  "Post-Annotation" = Counts1,
                  row.names = Studies,
                  check.names = FALSE
                  )


cols=colnames(Data)[-c(1)] # Columns with info
col_keys = colnames(Data) # Nombres unicos de las columnas
# Header
header = data.frame(col_keys = col_keys,
                    H1 = c(col_keys[c(1:4)], "Nº de genes", "Nº de genes"),
                    H2 = col_keys
                    )
#%>% width( width = 1.2, unit = "in")
#datatable(AnotResume, options = list(caption = "Summary  de la anotacion"))

```    

```{r echo=T, eval=doAnot, message=FALSE, warning=FALSE}
RX_table(Data, header = header, cols = cols, cols_keys = col_keys) %>%
  hline( border = fp_border(width = 1, color = "#878787"), part="body") %>%
  hline_bottom(border = fp_border(width = 2, color = "#007FA6"))

```

&nbsp;      


*Summary  anotaciones*  

```{r echo=T, eval=doAnot2, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=14, fig.cap=" _**Figure :** UpSet plot summary of study annotations._"}

ColorList = c("#cef587", "#90C432", "#2494b5","#03556e"  )
ColorPal <- colorRampPalette(ColorList)
Studies = c("E_MEXP_1425",
             "GSE2508",
             "GSE20950",
             "GSE29718", 
             "GSE64567", 
             "GSE92405",
             "GSE141432",
             "GSE205668")
Colors = ColorPal(length(Studies))
AnotDatas = glue("{Studies}_finalData")
MinSize = 10
An = sapply(AnotDatas, function(x) RX_ifelse(class(get(x))== "ExpressionSet",
                                             fData(get(x))$ENTREZID,
                                             rowData(get(x))$ENTREZID))
# Unique
IDs = unlist(An)
length(which(table(IDs) == 1))

An = list_to_matrix(An)
colnames(An) = Studies
Data = data.frame(An)

# Legend:
#Anots = as.data.frame(make_comb_mat(An)) 
#Leg = sapply(as.list(Anots), function(x) colnames(Data)[which(x == 1)], simplify = FALSE)
Inter = upset_data(Data,colnames(Data), min_size = MinSize)
Inter= str_split(names(Inter$sizes$exclusive_intersection), pattern = "-")
Legend = lapply(Inter, function(x) upset_query(intersect= x,
                                                color=Colors[length(x)],
                                                fill=Colors[length(x)],
                                                only_components=c('intersections_matrix', 'Intersection size')))

UpsetPlot <- upset(Data, colnames(Data), name = "Intersections", guides = "keep",
                   # Minimal number of annotation 
                   min_size = 10,
    width_ratio=0.16,
    stripes='transparent',
    matrix = intersection_matrix(
            geom=geom_point(
                shape='square',
                size=5,
                stroke = 1
            ), 
            segment=geom_segment(
              size= 1.2
            ),
            outline_color=list(
                active='black',
                inactive='gray'
            )) #+ theme(text=element_text(size = 29,angle=0))
    ,
    # Color asigmation
     queries= Legend, 
    # Set sizes
    set_sizes = upset_set_size(position='right') 
    # Number of counts
    + geom_text(aes(label=..count..), hjust=1.1, stat='count', color="white")
        + ylab('Number of annotated genes') + theme(axis.text.x=element_text(size = 10,
                                                                  angle=0),
                                               text = element_text(size=15),
                                         panel.grid.minor = element_blank(),
                                         panel.grid.major = element_blank(),
                                         axis.line.x = element_line(color="black"),
                                         axis.ticks = element_line(color="black")),
    # Interactions
    base_annotations = list(
        'Intersection size'=( 
            intersection_size(
              mapping=aes(fill='black'),
              text = list(
                vjust=0.5,
                hjust=-0.2,
                angle=90,
                size =4.5
            ), 
            text_colors=c(
                on_background='black', on_bar='white'
            )
            ) #+ scale_fill_manual(values=c('bars_color'='blue'))   
            + theme(panel.grid.major = element_blank(), 
               panel.grid.minor = element_blank(),
               plot.background = element_rect(fill="transparent",
                                                   color = "transparent"),
          panel.background = element_rect(fill = "transparent",
                                          color = "transparent"),
          axis.line.y = element_line(color="black"),
          axis.text.y=element_text(size = 10, angle=0)
          #axis.line.x = element_line(color="black"),
          #axis.ticks = element_line(color="black")
          )
            + ylab('Annotations per group')
        )), 
    themes=upset_modify_themes(
        list(
            'intersections_matrix'=theme(text=element_text(size=12)),
            'Intersection size'=theme(text=element_text(size=15)) )
        ) 
) 

# Saving...
ggsave(UpsetPlot, device = "svg", 
       filename = glue("{DirPlots}/UpsetPlot_norm_anot.svg"),height = 8, width = 12)

UpsetPlot
```  


```{r echo=F, eval=F, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=14}
# https://krassowski.github.io/complex-upset/articles/Examples_Python.html#themes
# https://krassowski.github.io/complex-upset/articles/Examples_R.html
```  

```{r echo=T, eval=doShow,  warning=FALSE, fig.align = centrado, fig.height=8, fig.width=12, fig.cap=" _**Figure :** UpSet plot summary of study annotations._"} 

knitr::include_graphics(glue("{DirPlots}/UpsetPlot_norm_anot.svg"))

```

&nbsp;   

***      

&nbsp;  


## Agrupation {.tabset .tabset-fade -}   


```{r echo=T, eval=loadData, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=14}
Studies = c("E_MEXP_1425",
             "GSE20950",
             "GSE29718", 
             "GSE64567", 
             "GSE92405",
             "GSE141432",
             "GSE205668") 

phenoDatas = lapply(glue("{Studies}_finalData"), 
                    function(x) data.frame(RX_ifelse(class(get(x)) == "SummarizedExperiment",
                                colData(get(x)),
                                pData(get(x))))[,c(1:9)]) 
names(phenoDatas) = Studies

pD = do.call(rbind, phenoDatas)
```


```{r echo=T, eval=loadData, message=FALSE, warning=FALSE, out.align = centrado, results = 'hide'}
# Summary table
Groups = c("C", "Ob", "T2D")
Tissues = levels(pD$Tissue)

DT = sapply( Tissues, function(Tissue)
  sapply(Studies, function(x)
    sapply(Groups, function(grupo)
      sum(unfactor(phenoDatas[[x]][which(phenoDatas[[x]]$Tissue == Tissue),"Group"]) == grupo),
      USE.NAMES = TRUE),
    USE.NAMES = TRUE),
  simplify = FALSE, USE.NAMES = TRUE)

DT2 = sapply(Tissues, function(i)
  data.frame(Tissue = i, Group = rownames(DT[[i]]), DT[[i]]),
  simplify = FALSE, USE.NAMES = TRUE)

DT3 = do.call("rbind", DT2) 

'datatable(DT3, rownames = NULL) %>% formatStyle("Tissue", target = "row", 
                                                backgroundColor = styleEqual(c("SAT", "VAT", "lSAT"), c('#A8DB47', '#E295F2', "#CFCFC8")))'
Data <- DT3 
```

```{r echo=T, eval=loadData, message=FALSE, warning=FALSE, out.align = centrado, out.width="98%"}
cols=colnames(Data)[-c(1)] # Columns with info
col_keys = colnames(Data) # Unique names
# Header
header = data.frame(col_keys = col_keys,
                    H1 = c(col_keys[c(1,2)], rep("Studies", length(Studies))),
                    H2 = col_keys
                    )
RX_table(Data, header = header, cols = cols, cols_keys = col_keys) %>%
  merge_v( j = c(1,2), part = "body") %>%
  hline(i=seq(from=3, to=nrow(Data)-1, by=3),
                border = fp_border(width = 1, color = "#878787"),
                part="body") %>%
  hline_bottom(border = fp_border(width = 2, color = "#007FA6"))

```
   



```{r echo=T, eval=doRes, message=FALSE, warning=FALSE, fig.align = centrado, fig.height=8, fig.width=12, fig.cap=" _**Figure :** Summary of the number of samples by group and sex comprising all studies._"}  

BarPlot <- RX_resumeBarPlot(data.frame(pD),
                            Fac_1 = "Group",
                            Fac_2 = "Sex",
                            Fac_wrap = "Tissue",
                            expandY= 0.1,
                            OneBar = TRUE,
                            colors = colors,
                            xlab = "Group",
                            Fac_1_tit = "Group",
                            Fac_2_tit = "Sexo",
                            sep_width = 0.8,
                            porcentLabels = FALSE,
                            alpha = alpha0)  

# Saving...
ggsave(BarPlot, device = "svg",
       filename = glue("{DirPlots}/ResumeBarPlot.svg"),
       height = 8, width = 12)

BarPlot

```  


```{r echo=T, eval=doShow,  warning=FALSE, fig.align = centrado, fig.height=5, fig.width=12, fig.cap=" _**Figure :** UpSet plot summary of study annotations._"} 

knitr::include_graphics(glue("{DirPlots}/ResumeBarPlot.svg"))

```



&nbsp;   

***      

&nbsp;  


